clients/java/src/main/java/io/camunda/zeebe/client/impl/http/CircularBufferInputStream.java

```diff
@@ -64,15 +64,27 @@ public synchronized void write(final ByteBuffer data) throws IOException {
     if (exception != null) {
       throw exception;
     }
-    if (data.limit() - data.position() > getAvailableSpace()) {
+    final int dataSize = data.remaining();
+    final int availableSpace = getAvailableSpace();
+    if (dataSize > availableSpace) {
       throw new IOException("Buffer is full");
     }
-    while (data.hasRemaining()) {
-      buffer[writePos] = data.get();
-      writePos = (writePos + 1) % capacity;
-      available++;
-      notifyAll(); // Notify any waiting readers
+
+    final int firstCopyLength = Math.min(dataSize, capacity - writePos);
+    data.get(buffer, writePos, firstCopyLength);
+    writePos = (writePos + firstCopyLength) % capacity;
+    availableData += firstCopyLength;
+
+    final int remaining = dataSize - firstCopyLength;
+    if (remaining > 0) {
+      // Wrap around
+      data.get(buffer, writePos, remaining);
+      writePos = (writePos + remaining) % capacity;
+      availableData += remaining;
     }
+
```

Comment:
🔧 If I'm not wrong `availableData += firstCopyLength;` and `availableData += remaining;` can be removed in favour of `availableData += dataSize;` just after this if statement